File: cpp/409.longest-palindrome.cpp
#include <string>
#include <vector>

using std::string;
using std::vector;

class Solution {
public:
  int longestPalindrome(string s) {
    int res = 0, mid = 0;
    const int N = 58;

    vector<int> cnt(N, 0);
    for (const char &c : s) {
      cnt[c - 'A']++;
    }

    for (int i = 0; i < N; ++i) {
      res += cnt[i] - cnt[i] % 2;
      mid |= cnt[i] % 2;
    }

    return res + mid;
  }
};

File: golang/409.longest-palindrome.go
const N = 'z' - 'A' + 1

func longestPalindrome(s string) int {
	cnt := make([]int, N)
	for i := range s {
		cnt[s[i]-'A']++
	}

	res, mid := 0, 0
	for _, v := range cnt {
		res += v - v%2
		mid |= v % 2
	}

	return res + mid
}
